control-plane/agents/src/bin/core/volume/specs.rs

```diff
@@ -271,30 +357,41 @@ pub(crate) async fn create_volume_replicas(
     registry: &Registry,
     request: &CreateVolume,
     volume: &VolumeSpec,
-) -> Result<CreateReplicaCandidate, SvcError> {
-    // Create a ag guard to prevent candidate collision.
-    let ag_guard = match registry.specs().get_or_create_affinity_group(volume) {
-        Some(ag) => Some(ag.operation_guard_wait().await?),
-        _ => None,
-    };
-
+) -> Result<Vec<CreateReplicaCandidate>, SvcError> {
     if !request.allowed_nodes().is_empty()
         && request.replicas > request.allowed_nodes().len() as u64
     {
         // oops, how would this even work mr requester?
         return Err(SvcError::InvalidArguments {});
     }
 
-    let node_replicas = volume_replica_candidates(registry, volume).await?;
+    let mut vec_create_replica_candidate: Vec<CreateReplicaCandidate> = Vec::new();
 
-    if request.replicas > node_replicas.len() as u64 {
-        Err(SvcError::from(NotEnough::OfPools {
-            have: node_replicas.len() as u64,
+    let node_replicas_vec: Vec<CreateReplicas> =
+        volume_replica_candidates_new(registry, volume).await?;
+
+    // The first element of the node_replicas_vec contains the most suitable replicas.
+    // Error out if the first element node_replicas_vec doesn't have enough replicas.
+    if request.replicas > node_replicas_vec[0].replicas().len() as u64 {
```

Comment:
you're assuming index  0 always exists, is this true?

Reply:
added more validation 
if node_replicas_vec.is_empty() {
        return Err(SvcError::from(NotEnough::OfPools {
            have: 0,
            need: request.replicas,
        }));
    }